internal/pool/pool.go

```diff
@@ -301,9 +301,9 @@ func (p *ConnPool) popIdle() *Conn {
 		return nil
 	}
 
-	idx := len(p.idleConns) - 1
-	cn := p.idleConns[idx]
-	p.idleConns = p.idleConns[:idx]
+	// Take idle conn from front
+	cn := p.idleConns[0]
+	p.idleConns = p.idleConns[1:]
```

Comment:
Each time you create a new slice with first element unaccessible any more. This is a memory leak until Go allocates new underlying storage time to time. It should be `p.idleConns = append(p.idleConns[:0], p.idleConns[1:]...)` but it is more expensive...

Also with such usage pattern connections don't become stale any more.

Reply:
I originally had `p.idleConns = append(p.idleConns[:0], p.idleConns[1:]...)` but changed it. If poolSize is 100 (as an example), then each popIdle will end up copying 100 pointers.

`p.idleConns = p.idleConns[1:]` will also reduce the capacity of the slice, and append will create a new array if the capacity is not sufficient. This is constant time in most cases. There shouldn't be a memory leak because the underlying array is eventually GCed when the array is doubled

Reply:
> Also with such usage pattern connections don't become stale any more so ReapStaleConns is not useful any more.

Yeah cycling through the connections makes this not very useful, unless there is very low traffic to redis